dfp/decimal/decimal

```diff
@@ -180,6 +180,286 @@ public:
 
 namespace std
 {
+
+  namespace
+  {
+    struct _Save_errno
+    {
+      _Save_errno() : _M_errno(errno) { errno = 0; }
+      ~_Save_errno() { if (errno == 0) errno = _M_errno; }
+      int _M_errno;
+    };
+
+    // calls setlocale(LC_ALL, "C") and returns a string containing the old
+    // locale name. Caller must delete[] the string. Returns NULL on error.
+    const char*
+    __set_C_locale()
+    {
+      char* __old = setlocale(LC_ALL, 0);
+      const size_t __len = strlen(__old) + 1;
+      char* __sav = new(nothrow) char[__len];
+      if (__sav)
+	{
+	  memcpy(__sav, __old, __len);
+	  setlocale(LC_ALL, "C");
+	}
+      return __sav;
+    }
+  }
+
+  template<>
+    void
+    __convert_to_v(const char* __s, decimal::decimal32& __v, ios_base::iostate& __err,
+		   const __c_locale&) throw()
+    {
+      // Assumes __s formatted for "C" locale.
+      const char* __sav = __set_C_locale();
+      if (!__sav)
+	{
+	  __err = ios_base::failbit;
+	  return;
+	}
+      char* __sanity;
+      bool __overflow = false;
+
+      __v = strtod32(__s, &__sanity);
+
+      // _GLIBCXX_RESOLVE_LIB_DEFECTS
+      // 23. Num_get overflow result.
+      if (__sanity == __s || *__sanity != '\0')
+	{
+	  __v = 0.0df;
+	  __err = ios_base::failbit;
+	}
+      else if (__overflow
+	       || __v == decimal::decimal32 {__builtin_infd32()}
+	       || __v == decimal::decimal32 {-__builtin_infd32()}
+	      )
+	{
+	  if (__v > decimal::decimal32 {0.0df})
+	    __v = DEC32_MAX;
+	  else
+	    __v = -DEC32_MAX;
+	  __err = ios_base::failbit;
+	}
+
+      setlocale(LC_ALL, __sav);
+      delete [] __sav;
+    }
+
+  template<>
+    void
+    __convert_to_v(const char* __s, decimal::decimal64& __v, ios_base::iostate& __err,
+		   const __c_locale&) throw()
+    {
+      // Assumes __s formatted for "C" locale.
+      const char* __sav = __set_C_locale();
+      if (!__sav)
+	{
+	  __err = ios_base::failbit;
+	  return;
+	}
+      char* __sanity;
+
+      __v = strtod64(__s, &__sanity);
+
+      // _GLIBCXX_RESOLVE_LIB_DEFECTS
+      // 23. Num_get overflow result.
+      if (__sanity == __s || *__sanity != '\0')
+	{
+	  __v = decimal::decimal64 {0.0dd};
+	  __err = ios_base::failbit;
+	}
+      else if (
+	       __v == decimal::decimal64 {__builtin_infd64()}
+	       || __v == decimal::decimal64 {-__builtin_infd64()} )
+	{
+	  if (__v > decimal::decimal64 {0.0dd})
+	    __v = DEC64_MAX;
+	  else
+	    __v = -DEC64_MAX;
+	  __err = ios_base::failbit;
+	}
+
+      setlocale(LC_ALL, __sav);
+      delete [] __sav;
+    }
+
+  template<>
+    void
+    __convert_to_v(const char* __s, decimal::decimal128& __v,
```

Comment:
Could these be parameterized based on the decimal type?

Reply:
These functions *are already* parameterized on the decimal type. But they are specialied for each  of `float`, `double` and `long double`, and that's because they have to do different things, using `strtof`, `strtod` or `strtold` as appropriate. You could have a more generic function which then uses some *other* function (or more likely, a traits class) to do the non-generic pieces that need to differ for each decimal type. For example, this one calls `strtod128` and uses `DEC128_MAX`, the specializations for `decimal32` and `decimal64` do not.

In the GCC code the differences are more significant, e.g. `strtof` might not be available, and `strtold` might be available but broken (on mingw, IIRC), so using a single generic function that handles all three types is harder. Here, that's not true. The functions are almost identical except for whether they call strtod32, strtod64, or strtod128, and which constants they use, so the variation could easily be hoisted out into a traits class. Those constants should come from `std::numeric_limits` anyway, reducing the differences between the specializations even further. For some reason `numeric_limits` isn't specialized in GCC's `<decimal/decimal>` header, but almost certainly should be.